`Cannot switch to the window` after/while interacting with child window

<!--
If you have all reproduction steps with a complete sample app, please share as many details as possible in the sections below.

Make sure that you tried using the latest TestCafe version (https://github.com/DevExpress/testcafe/releases), where this behavior might have been already addressed.

Before submitting an issue, please check CONTRIBUTING.md and existing issues in this repository (https://github.com/DevExpress/testcafe/issues) in case a similar issue exists or was already addressed. This may save your time (and ours).
-->

### What is your Test Scenario?
<!-- Describe what you'd like to test. -->
We have a number of tests which interact with the PayPal window, multiple times.

Example flow:
1. Make a Payment using PayPal
2. Return to Payment page, and make a 2nd Payment
3. Interact with PayPal 3rd time
4. Confirm payment successful
 
### What is the Current behavior?
<!-- Describe the behavior you see and consider invalid. -->
Although inconsistently, TestCafe often fails switching between windows; either the last step inside the paypal child window (click _confirm/continue_), or the 1st step after the window closes (_switching to the PayPal iFrame, to click again_): `Cannot switch to the window`.
For some reason the test for my demo app also fails with `A native alert dialog was invoked on page...` but that is unrelated since it's not happening on our actual tests.
 
### What is the Expected behavior?
<!-- Describe what you expected to happen. -->
 
### What is your web application and your TestCafe test code?
<!-- Share a public accessible link to your application or provide a simple app which we can run. -->
 
<details>
<summary>Your website URL (or attach your complete example):</summary>

<!-- Provide your website URL or attach a sample. Note: if your website requires any additional access procedures like authentication, please ask the website owner to send us a written confirmation at [support@devexpress.com](mailto:support@devexpress.com) in a free text form. It will allow the DevExpress staff to remotely access the website and its internal resources for research, testing, and debugging purposes. -->

If you check out my [testcafe_issues](https://github.com/rob4629/testcafe_issues) repo, you should just be able to use the test `Enter child window multiple times`
1. npm start -- starts the local server.
</details>

<details>
<summary>Your complete test code (or attach your test files):</summary>

<!-- Paste your test code here: --> 

```js
 fixture("Multiple Window Testing").page("localhost:4000")

const clickPayWithPaypal = async () => {
    const buttonIframe = Selector('#smart-button-container iframe');
    const payWithPaypalButton = Selector('[data-funding-source="paypal"]');

    await t
        .switchToIframe(buttonIframe)
        .expect(payWithPaypalButton.with({ visibilityCheck: true }).exists)
        .ok({ timeout: 40000 })
        .hover(payWithPaypalButton)
        .click(payWithPaypalButton)
        .switchToMainWindow();
}

const clickContinueButtonWithRetry = async () => {
    const acceptAllButton = Selector('#acceptAllButton');
    const continueButton = Selector('[data-testid="submit-button-initial"]');
    if (await acceptAllButton.visible) await t.click(acceptAllButton);
    await t.click(continueButton);
    if (await continueButton.visible) await t.click(continueButton);
  }

test("Enter child window multiple times", async () => {
    await clickPayWithPaypal();

    await t
      .typeText(Selector('#email'), <test-email>)
      .click(Selector('#btnNext'))
      .typeText(Selector('#password'), <test-password>)
      .click(Selector('#btnLogin'));
    await clickContinueButtonWithRetry();

    // Repeat
    await clickPayWithPaypal();
    await clickContinueButtonWithRetry();

    // Repeat
    await clickPayWithPaypal();
    await t.click(Selector('[data-testid="submit-button-initial"]')).wait(750);
});

```
</details>

<details>
<summary>Your complete configuration file (if any):</summary>

<!-- Paste your complete test config file here (even if it is huge): -->
```
{
  "assertionTimeout": 15000,
  "pageLoadTimeout": 30000,
  "selectorTimeout": 15000,

  "disableMultipleWindows": false,

  "quarantineMode": false,
  "skipJsErrors": true,

  "disablePageCaching": true,

  "reporter": [
    {
      "name": "xunit",
      "output": "artifacts/test-results/results.xml"
    },
    {
      "name": "spec"
    }
  ],

  "screenshots": {
    "fullPage": true,
    "takeOnFails": true,
    "pathPattern": "${TEST_ID}/${RUN_ID}_${FILE_INDEX}.png",
    "path": "artifacts/screenshots"
  }
}
```
</details>

<details>
<summary>Your complete test report:</summary>

<!-- Paste your complete result test report here (even if it is huge): -->
```
 Enter child window multiple times (screenshots:
 /Users/rcooper/Personal/testcafe_issues/artifacts/screenshots/test-1/_errors/1.png)

   1) A native alert dialog was invoked on page "http://localhost:4000/", but no handler was set for it. Use the
      "setNativeDialogHandler" function to introduce a handler function for native dialogs.

      Browser: Chrome 88.0.4324.150 / macOS 10.15.7


   2) Cannot switch to the window.

      Browser: Chrome 88.0.4324.150 / macOS 10.15.7
      Screenshot: null

         22 |}
         23 |
         24 |const clickContinueButtonWithRetry = async () => {
         25 |    const acceptAllButton = Selector('#acceptAllButton');
         26 |    const continueButton = Selector('[data-testid="submit-button-initial"]');
       > 27 |    if (await acceptAllButton.visible) await t.click(acceptAllButton);
         28 |    await t.setNativeDialogHandler(() => true).click(continueButton);
         29 |    if (await continueButton.visible) await t.setNativeDialogHandler(() => true).click(continueButton);
         30 |  }
         31 |
         32 |test("Click link within iFrame to open window", async () => {

         at <anonymous> (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:27:31)
         at asyncGeneratorStep (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:231)
         at _next (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:569)
         at <anonymous> (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:734)
         at <anonymous> (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:475)
         at clickContinueButtonWithRetry (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:24:35)
         at <anonymous> (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:58:11)
         at asyncGeneratorStep (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:231)
         at _next (/Users/rcooper/Personal/testcafe_issues/tests/multi_window_link.js:1:569)


 1/1 failed (45s)
npm ERR! Test failed.  See above for more details.
```
</details>

<details>
<summary>Screenshots:</summary>
<!-- If applicable, add screenshots to help explain the issue. -->

```

```
</details>

### Steps to Reproduce:
<!-- Describe what we should do to reproduce the behavior you encountered. -->

1. Check out my demo repo
2. Start the server (`npm start`)
3. Open a new console and run the `Enter child window multiple times` test
4. See the error...
 
### Your Environment details:
 
* testcafe version: 1.10.0-rc.4 > 1.10.2-alpha.2                  <!-- run `testcafe -v` -->
* node.js version: 12.14.3                   <!-- run `node -v` -->
* command-line arguments:  <!-- example: "testcafe ie,chrome -e test.js" -->
* browser name and version: Chrome 87 <!-- example: IE 11, Chrome 69, Firefox 100, etc. -->
* platform and version: Mac OS  <!-- example: "macOS 10.14, Windows, Linux Ubuntu 18.04.1, iOS 12 -->
* other:                                   <!-- any notes you consider important -->

--- a/src/client/driver/driver.js
+++ b/src/client/driver/driver.js
@@ -781,7 +781,10 @@ export default class Driver extends serviceUtils.EventEmitter {
 
         this._resolvePendingChildWindowInIframe();
 
-        this._onChildWindowOpened({ window: wnd, windowId: msg.windowId });
+        const childLinkExists = !!this.childWindowDriverLinks.find(link => link.windowId === msg.windowId);
+
+        if (!childLinkExists)
+            this._onChildWindowOpened({ window: wnd, windowId: msg.windowId });
     }
 
     _initChildDriverListening () {

--- a/src/client/driver/driver.js
+++ b/src/client/driver/driver.js
@@ -781,9 +781,9 @@ export default class Driver extends serviceUtils.EventEmitter {
 
         this._resolvePendingChildWindowInIframe();
 
-        const childLinkExists = !!this.childWindowDriverLinks.find(link => link.windowId === msg.windowId);
+        const childWindowDriverLinkExists = !!this.childWindowDriverLinks.find(link => link.windowId === msg.windowId);
 
-        if (!childLinkExists)
+        if (!childWindowDriverLinkExists)
             this._onChildWindowOpened({ window: wnd, windowId: msg.windowId });
     }
 

--- a/test/functional/fixtures/multiple-windows/pages/frame/frame.html
+++ b/test/functional/fixtures/multiple-windows/pages/frame/frame.html
@@ -13,7 +13,8 @@
     </script>
 </head>
 <body>
-    <a onclick="hideIframe()" href="./child.html" target="_blank">open window</a>
+    <a onclick="hideIframe()" href="./child.html" target="_blank">open window and hide iframe</a>
+    <a href="./child.html" target="_blank">open window</a>
 
     <button>frame</button>
 </body>

--- a/test/functional/fixtures/multiple-windows/test.js
+++ b/test/functional/fixtures/multiple-windows/test.js
@@ -304,7 +304,11 @@ describe('Multiple windows', () => {
 
     describe('iFrames', () => {
         it('Should switch to child window if it is opened in iFrame', () => {
-            return runTests('testcafe-fixtures/iframe.js', 'Open child window if iframe', { only: 'chrome' });
+            return runTests('testcafe-fixtures/iframe.js', 'Open child window if iframe');
+        });
+
+        it('Reload child window opened in iframe', () => {
+            return runTests('testcafe-fixtures/iframe.js', 'Reload child window opened in iframe');
         });
     });
 

--- a/test/functional/fixtures/multiple-windows/test.js
+++ b/test/functional/fixtures/multiple-windows/test.js
@@ -303,12 +303,12 @@ describe('Multiple windows', () => {
     });
 
     describe('iFrames', () => {
-        it('Should switch to child window if it is opened in iFrame', () => {
-            return runTests('testcafe-fixtures/iframe.js', 'Open child window if iframe');
+        it('Should switch to child window if it is opened from iFrame', () => {
+            return runTests('testcafe-fixtures/iframe.js', 'Open child window from iframe');
         });
 
-        it('Reload child window opened in iframe', () => {
-            return runTests('testcafe-fixtures/iframe.js', 'Reload child window opened in iframe');
+        it('Reload child window opened from iframe', () => {
+            return runTests('testcafe-fixtures/iframe.js', 'Reload child window opened from iframe');
         });
     });
 

--- a/test/functional/fixtures/multiple-windows/testcafe-fixtures/iframe.js
+++ b/test/functional/fixtures/multiple-windows/testcafe-fixtures/iframe.js
@@ -1,4 +1,7 @@
-import { Selector } from 'testcafe';
+import { Selector, ClientFunction } from 'testcafe';
+
+const reload = ClientFunction(() => window.location.reload());
+const close  = ClientFunction(() => window.close());
 
 fixture `iFrames`
     .page `http://localhost:3000/fixtures/multiple-windows/pages/frame/parent.html`;
@@ -8,7 +11,7 @@ test('Open child window if iframe', async t => {
 
     await t.click(Selector('button').withText('frame'));
 
-    await t.click('a');
+    await t.click(Selector('a').withExactText('open window and hide iframe'));
 
     await t.click(Selector('button').withText('child'));
 
@@ -19,4 +22,21 @@ test('Open child window if iframe', async t => {
     await t.switchToPreviousWindow();
 
     await t.click(Selector('button').withText('child'));
+});
+
+test('Reload child window opened in iframe', async t => {
+    await t.switchToIframe('iframe');
+
+    await t.click(Selector('a').withExactText('open window'));
+
+    for (let i = 0; i < 10; i++)
+        await reload();
+
+    await close();
+
+    await t.switchToIframe('iframe');
+
+    await t.click('a');
+
+    await close();
 });

--- a/test/functional/fixtures/multiple-windows/testcafe-fixtures/iframe.js
+++ b/test/functional/fixtures/multiple-windows/testcafe-fixtures/iframe.js
@@ -6,7 +6,7 @@ const close  = ClientFunction(() => window.close());
 fixture `iFrames`
     .page `http://localhost:3000/fixtures/multiple-windows/pages/frame/parent.html`;
 
-test('Open child window if iframe', async t => {
+test('Open child window from iframe', async t => {
     await t.switchToIframe('iframe');
 
     await t.click(Selector('button').withText('frame'));
@@ -24,7 +24,7 @@ test('Open child window if iframe', async t => {
     await t.click(Selector('button').withText('child'));
 });
 
-test('Reload child window opened in iframe', async t => {
+test('Reload child window opened from iframe', async t => {
     await t.switchToIframe('iframe');
 
     await t.click(Selector('a').withExactText('open window'));

